Add linear solver namespace and IterationId

nilsvu opened the pull request:
## Proposed changes

This is in preparation for the linear solver algorithms that I will add in future PRs. In a way, the linear solver is for elliptic systems what time stepping is for the evolution code. The elliptic solver needs to solve linear systems of equations of the type $Ax=b$, where $A$ is a global matrix representing the DG discretisation of the problem to solve. We never construct the full matrix $A$ but only need the projection $Ax$ that can be evaluated locally by virtue of the DG formulation.

The algorithms that I will add in the near future are a conjugate gradient method for symmetric operators (#792) and a GMRES method for non-symmetric operators.  

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [x] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [x] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [x] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

nilsvu committed:
Add linear solver namespace and IterationId

wthrowe commented on src/DataStructures/DenseVector.hpp:
```diff
@@ -10,6 +10,7 @@
 
 #include "Utilities/Blaze.hpp"
 #include "Utilities/TypeTraits.hpp"
+#include "Utilities/MakeWithValue.hpp"
```
Alphabetize.

nilsvu replied:
I removed this commit for now and will prepend it to the CG algorithm PR, since it has nothing to do with this PR

wthrowe commented on src/NumericalAlgorithms/LinearSolver/IterationId.hpp:
```diff
@@ -0,0 +1,49 @@
+
+  IterationId() = default;
+
+  IterationId(const size_t step_number) noexcept : step_number(step_number) {};
```
No `;`.

wthrowe commented on src/NumericalAlgorithms/LinearSolver/Tags.hpp:
```diff
@@ -0,0 +1,23 @@
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines DataBox tags for scalar wave system
```
Fix.

wthrowe reviewed (changes requested):
Based on the PR title I was expecting documentation for the LinearSolver namespace.  Is it intentional that that is not here?

nilsvu commented:
@wthrowe I added a brief documentation for the namespace. Should I actually create a Doxygen group, too?
**Edit:** Also added a group and more documentation. Please let me know if this suffices for now.

kidder commented on src/NumericalAlgorithms/LinearSolver/IterationId.hpp:
```diff
@@ -0,0 +1,53 @@
+#include <functional>
+#include <iosfwd>
+
+namespace PUP {
```
/// \cond

kidder commented on src/NumericalAlgorithms/LinearSolver/IterationId.hpp:
```diff
@@ -0,0 +1,53 @@
+namespace PUP {
+class er;
+}  // namespace PUP
+
```
/// \endcond

kidder reviewed (changes requested):
Also test the names of the other tags

nilsvu commented:
This should be ready

wthrowe commented on docs/GroupDefs.hpp:
```diff
@@ -453,6 +453,35 @@
+ * the computational domain it becomes unfeasible to numerically invert the
+ * global matrix \f$A\f$. Instead, we solve the problem iteratively so that we
+ * never need to construct \f$A\f$ globally but only need the projection
+ * \f$Ax\f$ that can be evaluated locally by virtue of the DG formulation. This
```
This may just be a terminology difference between fields, but I am confused by this being referred to as a projection, since in the linear algebra sense "Ax" is not a projection of either "A" or "x".

nilsvu replied:
Agreed. Do you have a suggestion for a better name?

I mean to say "The linear operator applied to the `Tag`". The wrapped tag will typically be the physical fields Variables, or the `LinearSolver::Tags::Operand` which takes the role of the evolved Variables. In this analogy, the linear operator applied to the operand takes the role of the `Tags::dt` prefix.

wthrowe replied:
The tag should also be renamed, but at the moment I'm confused by the documentation, since I'm not sure what "Ax" refers to here.  Is it the global A times the global x, some local A times the local x?  Is it the first of those restricted to the element?  Something else?

nilsvu replied:
Yes, the first of those restricted to the element. That’s what I mean with the comment that it can be evaluated locally by virtue of the DG formulation.

nilsvu replied:
In terms of names, all that I can think of is along the lines of:

- `LinearSolver::Tags::op<Tag>`
- `LinearSolver::Tags::Operator<Tag>`
- `LinearSolver::Tags::OperatorValue<Tag>`
- `LinearSolver::Tags::OperatorAction<Tag>`
- `LinearSolver::Tags::OperatorAppliedTo<Tag>`

Perhaps `Operator<Tag>` is fine because it closely matches `A(x)`?

wthrowe commented on src/NumericalAlgorithms/LinearSolver/IterationId.hpp:
```diff
@@ -0,0 +1,55 @@
+
+  IterationId() = default;
+
+  explicit IterationId(const size_t initial_step_number) noexcept
```
`initial_step_number` suggests the start of the run to me (presumably that is always step 0).  I believe I've seen an `_in` suffix used for this, but here I'd just leave the constructor out and let people use aggregate initialization.

wthrowe commented on src/NumericalAlgorithms/LinearSolver/Tags.hpp:
```diff
@@ -0,0 +1,89 @@
+
+/*!
+ * \ingroup LinearSolverGroup
+ * \brief The residual \f$r=b - Ax\f$
```
Where does `b` come from in all of this?

nilsvu replied:
If you mean where is it defined: In the documentation where I write $Ax=b$ is the linear equation to be solved.
If you mean how is it computed: In the element initializations from the problem at hand.

wthrowe replied:
I mean where is it stored or otherwise obtained from when setting this tag.

nilsvu replied:
During initialization it’s computed and passed to the linear solver init action. It typically doesn’t need to be stored, since the residual is updated iteratively.

nilsvu replied:
Oh and where it's handled it's represented by the existing `Tags::Source` prefix.

wthrowe commented on tests/Unit/NumericalAlgorithms/LinearSolver/Test_LinearSolverTags.cpp:
```diff
@@ -0,0 +1,25 @@
+// Distributed under the MIT License.
```
Match the name of the source file, i.e., `Test_Tags.cpp`.

nilsdeppe reviewed (dismissed):
Dox rendered: http://gh.nilsdeppe.com/spectre/pr_854/group__LinearSolverGroup.html

kidder reviewed (dismissed):
@wthrowe please look at; I am happy for this to be squashed

nilsvu commented:
@nilsdeppe @kidder could you comment on the name of the operator tag that @wthrowe and I discussed above, if you have anything to say about it?

kidder commented:
I prefer OperatorAppliedTo

nilsdeppe commented:
I'm fine with what @kidder  said, don't really care. Whatever means we can merge this sooner...

nilsvu commented:
squashed and rebased

nilsvu commented:
The tale continues in #867 :)

kidder commented:
@wthrowe please look at

kidder commented:
travis failures are timeouts
